### ISSUE TITLE
Keras TF ImageNet Preprocessing contradicts TF, may affect pretrained weights

### ISSUE THREAD

**Issue by ahundt** [keras-team/keras#8916]
It appears Keras' imagenet image preprocessing may be inconsistent with how it is done in tf, in particular keras sets values to `[-1, 1]` while in tf the expected range is `[0, 1]`.

- [slim/preprocessing/inception_preprocessing.py](https://github.com/tensorflow/models/blob/master/research/slim/preprocessing/inception_preprocessing.py#L284) 
    - notes `If dtype is tf.float32 then the range should be [0, 1]`
- [keras preprocess_input()](https://github.com/keras-team/keras/blob/master/keras/applications/imagenet_utils.py#L72) 
    - notes `tf: will scale pixels between -1 and 1, sample-wise`.

Here is the key doc from Keras' `preprocess_input()`:

```python

def preprocess_input(x, data_format=None, mode='caffe'):
    """Preprocesses a tensor encoding a batch of images.
    # Arguments
        x: input Numpy or symoblic tensor, 3D or 4D.
        data_format: data format of the image tensor.
        mode: One of "caffe", "tf".
            - caffe: will convert the images from RGB to BGR,
                then will zero-center each color channel with
                respect to the ImageNet dataset,
                without scaling.
            - tf: will scale pixels between -1 and 1,
                sample-wise.
    # Returns
        Preprocessed tensor.
    """
```

Here are the key docs from the tf slim function `preprocess_for_train()` which specify a [0, 1] range:

```python
def preprocess_for_train(image, height, width, bbox,
                         fast_mode=True,
                         scope=None,
                         add_image_summaries=True):
  """Distort one image for training a network.
  Distorting images provides a useful technique for augmenting the data
  set during training in order to make the network invariant to aspects
  of the image that do not effect the label.
  Additionally it would create image_summaries to display the different
  transformations applied to the image.
  Args:
    image: 3-D Tensor of image. If dtype is tf.float32 then the range should be
      [0, 1], otherwise it would converted to tf.float32 assuming that the range
      is [0, MAX], where MAX is largest positive representable number for
      int(8/16/32) data type (see `tf.image.convert_image_dtype` for details).
    height: integer
    width: integer
    bbox: 3-D float Tensor of bounding boxes arranged [1, num_boxes, coords]
      where each coordinate is [0, 1) and the coordinates are arranged
      as [ymin, xmin, ymax, xmax].
    fast_mode: Optional boolean, if True avoids slower transformations (i.e.
      bi-cubic resizing, random_hue or random_contrast).
    scope: Optional scope for name_scope.
    add_image_summaries: Enable image summaries.
  Returns:
    3-D float Tensor of distorted image used for training with range [-1, 1].
  """
```

side notes: 

- https://github.com/tensorflow/tensorflow/issues/15722 is the corresponding tf issue 
- the [tf image adjustments guide](https://www.tensorflow.org/api_guides/python/image#Image_Adjustments) doesn't have much documenting the inputs.

**Comment by myutwo150** [comment]
The `[0, 1]` range is the input to `preprocess_for_train`, the output range is `[-1, 1]`. They are the same.

**Comment by ahundt** [comment]
My comments from https://github.com/tensorflow/tensorflow/issues/15722#issuecomment-354625196:

I believe I've got it working, there isn't a bug, but the APIs were simply not clear about the preconditions and postconditions of each function. Here is a summary, please correct me if I misunderstood:

[tf image adjustments guide](https://www.tensorflow.org/api_guides/python/image#Image_Adjustments) APIs:

1. Given raw RGB image with values in the range `[0, 255]` 
2. call [tf.image.per_image_standardization()](https://www.tensorflow.org/api_docs/python/tf/image/per_image_standardization), output values are in `[0, 1]` range.
3. expected input range of the image augmentation functions is `[0, 1]`, they can be called at this point if training, skip otherwise
4. subtract 0.5 from that output and multiply by 2 to get a range of `[-1, 1]`
5. The expected input range when you actually feed the data to the various pretrained neural networks in the model zoo is `[-1, 1]`.

for [keras preprocess_input()](https://github.com/keras-team/keras/blob/master/keras/applications/imagenet_utils.py#L72) :

1. Input is a raw RGB 0-255 image and output is `[-1, 1]`
    - if it is inference time this can be fed directly to networks and in that case you are done.
2. If you wanted to utilize tf augmentation you must divide by 2 and add 0.5, to put values in `[0, 1]` range.
3. apply the augmentation from [tf image adjustments guide](https://www.tensorflow.org/api_guides/python/image#Image_Adjustments) APIs
4. subtract 0.5 from that output and multiply by 2 to get a range of `[-1, 1]`
5. perform training step


Perhaps this can be noted clearly in these APIs?

**Comment by ztf-ucas** [comment]
@ahundt 
When I use the pretrained vgg16 model(vgg16_weights_tf_dim_ordering_tf_kernels), Which way should I according to preprocess input? Tf or caffe?
Keras2.1.2 tf backend
The pretrained model are downloaded from weights_path_no_top in Keras code.

**Comment by ahundt** [comment]
@ztf-ucas I ran lots of automated experiments with both to try and find an optimal choice, and I recall it honestly didn't make a big difference in my use case. I simply default to 'tf' now, but as with all neural networks your mileage may vary based on your dataset and problem type.

**Comment by ztf-ucas** [comment]
> @ztf-ucas I ran lots of automated experiments with both to try and find an optimal choice, and I recall it honestly didn't make a big difference in my use case. I simply default to 'tf' now, but as with all neural networks your mileage may vary based on your dataset and problem type.

Thank you very much!!!

**Comment by ymodak** [comment]
Closing this issue since its resolved. Feel free to reopen if the problem still persists. Thanks!

**Comment by ahundt** [comment]
@ymodak where was this resolved? #9273 did not resolve it.

**Comment by ymodak** [comment]
Yes you are right. Thanks for catching this. Reopening the issue since its still remains.
